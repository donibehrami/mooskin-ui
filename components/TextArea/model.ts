import { ITextAreaBoxComponentProps } from '../Box/model';
import { IInputCallbackData } from '../index';

export interface ITextAreaComponentProps extends ITextAreaBoxComponentProps {
    /** what data is being used, helps whn extracting user input, you know on what field changes are made */
    dataLabel?: string;

    /** callback that is called when the textarea changes */
    onChangeTextArea: (e: React.ChangeEvent<HTMLTextAreaElement>, data: IInputCallbackData) => void;
}
