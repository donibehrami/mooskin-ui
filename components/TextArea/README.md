# MooSkin UI - TextArea Component

The MooSkin TextArea Component works similar to the normal HTML `<textarea></textarea>` element, but with pre-defined styling and attributes.

___

### Usage

To start using the TextArea Component first you have to Import it

```
Import {TextArea} from 'mooskin';
```
or modular import
```
// JS
import TextArea from 'mooskin/lib/TextArea';

// CSS
import 'mooskin/lib/TextArea/style.css';
```

And then you can simply start using it by typing

```
<TextArea />
```

For easy use, the components are named similar to normal HTML components but with a capital first letter.

Like the ```<textarea/>``` element it will accept given attributes and render differently based on the given attributes

### Examples

Passing the placeholder attribute

```
<TextArea placeholder="username" />
```

or making it disabled with a value, in this case the proper disabled class will be loaded (ex. disabling cursor)

```
<TextArea value="This one is disabled" disabled />
```

or pass it a function for event handling

```
<TextArea onChange={yourFunc} />
```

or just give it a custom style

```
<TextArea style={yourStyle} />
```

### Callback

The TextArea Component Callback will always return the value of the TextArea on each change. This can be used with a function passed via the `onChange` prop. For example, if u want to console log the callback value, pass a function to the `onChange` prop.

```
const logValue = (e, data) => {     // data is the callback object, which consists of value and a dataLabel(not required)
    console.log(data.value);
};

<TextArea onChange={logValue} />
```
In this case on each TextArea change, the value will be console logged.

This can be used in various situations and combinations, for an enhanced development experience.

### Validation

The Textarea Component can be validated using the `status` and `validate` props. The `status` prop will accept an 'error' or 'success' string to apply the appropriate classes to the textarea, or none for that matter. The `validate` prop is the validation function which accepts a arg of an object containing `value`, `dataLabel` and `required` props of the component as properties, and depending on the received value, you can use any kind of validation you like. The `description` prop can also be used in the process to inform the user about validation errors. The validation will fire the first time onBlur, and every other time onChange.

#### Example

```
this.state = {
    status: '',
    message: '',
    value: ''
}

<TextArea
    value={this.state.value}
    validate={this.validate}
    status={this.state.status}
    description={this.state.message}
    onChange={this.onChange}
/>

onChange(e, data) {
    this.setState({value: data.value})
}

validate(data){
    if (data.value){
        if (data.value.length < 5){
            this.setState({status: 'error', message: 'TextArea should have 5 or more characters'});
        } else {
            this.setState({status: '', message: ''})
        }
    } else {
        this.setState({status: 'error', message: 'This TextArea is required'});
    }
}
```

In this case, when the user focuses the textarea and blurs away immediately, an error class will get applied to the textarea and the user will get informed that thee textarea is required. When the user starts typing, the error message will change, saying that the textarea should have 5 or more characters, when this is done, the error classes will get removed. Alternatevely, when the validation passes, we can pass a success status to apply the related class.

<div class="playground-doc">

## Supported attributes

* `id` - id of the element
* `type` - type of the textarea (eg. text)
* `name` - name of the textarea
* `value` - text value of textarea
* `placeholder` - backtext description of textarea
* `minlength` - min number of characters allowed
* `maxlength` - max number of characters allowed
* `disabled` - textarea field should be disabled
* `required` - textarea field is required
* `description` - textarea description (small italic bottom)
* `spacing` - spacing between label and textarea
* `readonly` - make textarea readonly
* `className` - css class
* `dataLabel` - label what kind of data 
* `style` - textarea field style
* `onChange` - callback to be triggered on textarea change
* `validate` - validate function
* `status` - to be used with validate, to apply classes based on error or success

</div>

Allthough these attributes are supported, all of them are optional. 
NOTE! While in rich editor mode only `richValue` and `onEditorChange` props will take effect, and vice versa.

#### For more

___

[MooSkin-UI](https://github.com/moosend/mooskin-ui)