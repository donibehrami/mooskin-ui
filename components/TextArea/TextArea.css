@import '../_utils/globals/variables.css';
@import '../_utils/globals/screens.css';

.areaContainer{
    display: flex;
}

.textAreaDiv{
    width: 100%;
}

.textarea{
    padding: 7px;
    font-family: Arial,sans-serif;
    border: 0;
    color: #9e9e9e;
    font-size: 16px;
    border-radius: 8px;
    background: #efefef;
    display: table-cell;
    width: 100%;
}

.textarea::-webkit-scrollbar {
    width: 5px;
    height: 10px;
}

.textarea::-webkit-scrollbar-button:start:decrement,
.textarea::-webkit-scrollbar-button:end:increment  {
    display: none;
}

.textarea::-webkit-scrollbar-track-piece  {
    background-color: #eeeeee;
}

textarea::-webkit-scrollbar-thumb:vertical {
    background: #d9d9d9;
    border-radius: 16px;
}

.disabledTextArea{
    cursor:not-allowed;
    color: #bfbfbf;
    background: #f6f6f6;
    resize: none;
}

.textarea::placeholder {
    color: var(--inputPlaceholder);
}

.textarea:focus {
    outline: none;
    color:#333333;
}

.textareaLabel {
    color: var(--inputLabel);
    font-size: 13px;
    font-size: .92857142857rem;
    padding-top: 7px;
    cursor: pointer;
    width: 155px;
}

.areaContainer .description{
    display: block;
    color: #b3b3b3;
    font-size: .78571428571rem;
    margin-left: 3px;
}

@media (--xSmall){
    .areaContainer{display: flex; flex-direction: column}
    .textareaLabel{padding-top: 0px; margin-left: 3px}
}