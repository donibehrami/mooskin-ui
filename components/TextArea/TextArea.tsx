import * as React from 'react';

// Helpers
// import { getBoxProps } from '../_utils/helper';

// Models
import { IDescriptionComponentProps } from '../Description/model';
import { ILabelComponentProps } from '../Label/model';
import { ITextAreaComponentProps } from './model';

// Components
import { Description } from '../Description/Description';
import { Label } from '../Label/Label';

// Styled Components
// import {StyledTextArea} from './styles';

export const TextArea: React.FC<ITextAreaComponentProps> = (props) => {

    // const onChange = (e: React.ChangeEvent<HTMLTextAreaElement>) => {
    //     !props.disabled && props.onChangeTextarea &&
    //     props.onChangeTextarea(e, {value: e.target.value, dataLabel: props.dataLabel});

    //     !props.disabled && props.onChange && props.onChange(e);
    // };

    // return <StyledTextArea{...props} boxAs="textarea" onChange={onChange} />;
    return null;
};

TextArea.defaultProps = {
    className: '',
    style: {}
};

TextArea.displayName = 'TextArea';

export default TextArea;

/**
 * TextareaLabel
 */
export const TextareaLabel: React.FC<ILabelComponentProps> = (props) => {
    return <Label disabled={props.disabled} {...props} />;
};

TextareaLabel.defaultProps = {
    className: '',
    style: {}
};

TextareaLabel.displayName = 'TextareaLabel';

/**
 * TextareaDescription
 */
export const TextareaDescription: React.FC<IDescriptionComponentProps> = (props) => {
    return <Description {...props} />;
};

TextareaDescription.defaultProps = {
    className: '',
    style: {}
};

TextareaDescription.displayName = 'TextareaDescription';
