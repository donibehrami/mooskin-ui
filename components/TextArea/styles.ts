import styled from 'styled-components';

// Models
import { ITextAreaBoxComponentProps } from '../Box/model';

// Components
import { Box } from '../Box/Box';

// "CSS" variables
import variables from '../_utils/globals/variables';

export const StyledTextArea = styled(Box)<ITextAreaBoxComponentProps>`
    padding: 11px;
    border-radius: 3px;
    color: ${(props) => props.palette?.fontColors.common || variables.fontColors.common};
    border: solid 1px ${(props) => props.palette?.borderColors.common || variables.borderColors.common};
    background-color: ${(props) => props.palette?.backgroundColors.common || variables.backgroundColors.common};
    opacity: ${(props) => (props.disabled ? 0.5 : 1)};
    font-family: Hind;
    font-size: 12px;
`;

StyledTextArea.displayName = 'StyledTextArea';
