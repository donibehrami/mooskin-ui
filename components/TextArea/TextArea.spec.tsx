import * as React from 'react';
import TextArea from './TextArea';

import { shallow } from 'enzyme';

describe('TextArea', () => {
    test('renders correctly', () => {
        const func = jest.fn();

        const tree = shallow(<TextArea value="asd" onChange={func} />);
        expect(tree).toMatchSnapshot();
    });

    test('renders properly into dom and has Placeholder prop', () => {
        const component = shallow(<TextArea value="value" placeholder="username" />);

        expect(component.find('StyledTextArea').prop('placeholder')).toContain('username');
    });

    test('renders an input with a "required" prop and minlength', () => {
        const component = shallow(<TextArea value="value" minLength={5} required />);

        expect(component.find('StyledTextArea').prop('required')).toEqual(true);
        expect(component.find('StyledTextArea').prop('minLength')).toEqual(5);
    });

    test('renders an input with a passed value and maxlength', () => {
        const component = shallow(<TextArea value="random" maxLength={50} />);

        expect(component.find('StyledTextArea').prop('value')).toEqual('random');
        expect(component.find('StyledTextArea').prop('maxLength')).toEqual(50);
    });

    test('renders an input with custom css class and style', () => {
        const component = shallow(<TextArea value="value" style={{ color: 'blue' }} className="input-group" />);

        expect(component.find('StyledTextArea').hasClass('input-group')).toBe(true);
        expect(component.find('StyledTextArea').prop('style')).toEqual({ color: 'blue' });
    });

    test('onChange prop callback is called when a key is pressed', () => {
        const func = jest.fn();

        const component = shallow(<TextArea value="value" onChange={func} />);

        component.find('StyledTextArea').simulate('change', { target: { value: 'text' } });
        expect(func).toHaveBeenCalled();
    });
});
