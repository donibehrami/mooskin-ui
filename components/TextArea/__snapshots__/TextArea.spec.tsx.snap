// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TextArea renders correctly 1`] = `
<TextArea
  className=""
  onChange={[MockFunction]}
  style={Object {}}
  value="asd"
>
  <Component
    className=""
    onChange={[MockFunction]}
    palette={
      Object {
        "backgroundColors": Object {
          "button": "#3fbaca",
          "common": "#ffffff",
          "disabledButton": "#e2e2e2",
          "main": "#f7f7f7",
          "secondaryToggle": "#ffffff",
          "toggle": "#425579",
        },
        "borderColors": Object {
          "common": "#e2e2e2",
        },
        "fontColors": Object {
          "anchor": "#3fbaca",
          "button": "#ffffff",
          "common": "#293346",
          "description": "#9d9d9d",
          "disabledAnchor": "#9d9d9d",
          "label": "#2d2d2d",
          "secondaryToggle": "#9d9d9d",
          "text": "#2d2d2d",
          "toggle": "#ffffff",
        },
      }
    }
    style={Object {}}
    value="asd"
  >
    <StyledTextArea
      boxAs="textarea"
      className=""
      onChange={[Function]}
      palette={
        Object {
          "backgroundColors": Object {
            "button": "#3fbaca",
            "common": "#ffffff",
            "disabledButton": "#e2e2e2",
            "main": "#f7f7f7",
            "secondaryToggle": "#ffffff",
            "toggle": "#425579",
          },
          "borderColors": Object {
            "common": "#e2e2e2",
          },
          "fontColors": Object {
            "anchor": "#3fbaca",
            "button": "#ffffff",
            "common": "#293346",
            "description": "#9d9d9d",
            "disabledAnchor": "#9d9d9d",
            "label": "#2d2d2d",
            "secondaryToggle": "#9d9d9d",
            "text": "#2d2d2d",
            "toggle": "#ffffff",
          },
        }
      }
      style={Object {}}
      value="asd"
    >
      <Box
        boxAs="textarea"
        className="sc-gtssRu zOXFd"
        onChange={[Function]}
        palette={
          Object {
            "backgroundColors": Object {
              "button": "#3fbaca",
              "common": "#ffffff",
              "disabledButton": "#e2e2e2",
              "main": "#f7f7f7",
              "secondaryToggle": "#ffffff",
              "toggle": "#425579",
            },
            "borderColors": Object {
              "common": "#e2e2e2",
            },
            "fontColors": Object {
              "anchor": "#3fbaca",
              "button": "#ffffff",
              "common": "#293346",
              "description": "#9d9d9d",
              "disabledAnchor": "#9d9d9d",
              "label": "#2d2d2d",
              "secondaryToggle": "#9d9d9d",
              "text": "#2d2d2d",
              "toggle": "#ffffff",
            },
          }
        }
        style={Object {}}
        value="asd"
      >
        <Component
          boxAs="textarea"
          className="sc-gtssRu zOXFd"
          onChange={[Function]}
          palette={
            Object {
              "backgroundColors": Object {
                "button": "#3fbaca",
                "common": "#ffffff",
                "disabledButton": "#e2e2e2",
                "main": "#f7f7f7",
                "secondaryToggle": "#ffffff",
                "toggle": "#425579",
              },
              "borderColors": Object {
                "common": "#e2e2e2",
              },
              "fontColors": Object {
                "anchor": "#3fbaca",
                "button": "#ffffff",
                "common": "#293346",
                "description": "#9d9d9d",
                "disabledAnchor": "#9d9d9d",
                "label": "#2d2d2d",
                "secondaryToggle": "#9d9d9d",
                "text": "#2d2d2d",
                "toggle": "#ffffff",
              },
            }
          }
          style={Object {}}
          value="asd"
        >
          <StyledBox
            as="textarea"
            boxAs="textarea"
            className="sc-gtssRu zOXFd"
            onChange={[Function]}
            palette={
              Object {
                "backgroundColors": Object {
                  "button": "#3fbaca",
                  "common": "#ffffff",
                  "disabledButton": "#e2e2e2",
                  "main": "#f7f7f7",
                  "secondaryToggle": "#ffffff",
                  "toggle": "#425579",
                },
                "borderColors": Object {
                  "common": "#e2e2e2",
                },
                "fontColors": Object {
                  "anchor": "#3fbaca",
                  "button": "#ffffff",
                  "common": "#293346",
                  "description": "#9d9d9d",
                  "disabledAnchor": "#9d9d9d",
                  "label": "#2d2d2d",
                  "secondaryToggle": "#9d9d9d",
                  "text": "#2d2d2d",
                  "toggle": "#ffffff",
                },
              }
            }
            style={Object {}}
            value="asd"
          >
            <textarea
              className="sc-bdnylx hKVIpd sc-gtssRu zOXFd"
              onChange={[Function]}
              style={Object {}}
              value="asd"
            />
          </StyledBox>
        </Component>
      </Box>
    </StyledTextArea>
  </Component>
</TextArea>
`;
